In [1]:
x = 12
x

12

In [2]:
type(x)

int

In [3]:
y = 34

In [4]:
py_sum = x + y
py_sum

46

In [5]:
type(py_sum)

int

In [6]:
x_flot = 12.0

In [7]:
sum_flot_int = x_flot + y
sum_flot_int

46.0

In [8]:
type(sum_flot_int)

float

### Scalars in PyTorch

In [58]:
import torch as pt

In [60]:
x_pt = pt.tensor(25) 
x_pt

tensor(25)

In [61]:
x_pt.shape

torch.Size([])

### Scalars in TensorFlow 

In [95]:
import tensorflow as tf

In [13]:
x_tf = tf.Variable (3, dtype = tf.int16)
x_tf

<tf.Variable 'Variable:0' shape=() dtype=int16, numpy=3>

In [19]:
x_tf.shape

TensorShape([])

In [22]:
y_tf = tf.Variable (22, dtype = tf.int16)
y_tf

<tf.Variable 'Variable:0' shape=() dtype=int16, numpy=22>

In [24]:
sum_x_y_tf = tf.add(x_tf,y_tf) # same as x_tf + y_tf
sum_x_y_tf

<tf.Tensor: shape=(), dtype=int16, numpy=25>

In [26]:
sum_x_y_tf.numpy() # numpy operations automatically convert tensor to numpy and vice versa. 

25

In [27]:
type(sum_x_y_tf.numpy())

numpy.int16

### Vectors in numpy

In [2]:
import numpy as np

In [3]:
x = np.array([25,3,5])
x

array([25,  3,  5])

In [5]:
len(x)

3

In [6]:
x.shape

(3,)

In [7]:
type(x)

numpy.ndarray

In [8]:
x[0]

25

In [11]:
x[2]

5

### Vector Transposition 

In [12]:
x_t = x.T

In [13]:
x_t

array([25,  3,  5])

In [14]:
x_t.shape

(3,)

In [17]:
y = np.array([[2,3,5]])
y

array([[2, 3, 5]])

In [18]:
y.shape

(1, 3)

In [19]:
y_t = y.T
y_t

array([[2],
       [3],
       [5]])

In [20]:
y_t.shape

(3, 1)

### zero vectors

In [21]:
z = np.zeros(4)
z

array([0., 0., 0., 0.])

###  L^2  Norm

In [22]:
x

array([25,  3,  5])

In [25]:
( (25**2) + (3**2) + (5**2) ) ** (1/2)

25.67099530598687

###### It means the length of the vector is 25.67 unit.  

In [24]:
np.linalg.norm(x)

25.67099530598687

### L^1 Norm

In [26]:
x

array([25,  3,  5])

In [27]:
np.abs(x[0]) + np.abs(x[1]) + np.abs(x[2])

33

### Squared L^2 Norm

In [28]:
x

array([25,  3,  5])

In [29]:
np.abs(x[0])**2 + np.abs(x[1])**2 + np.abs(x[2])**2

659

In [30]:
np.dot(x, x)

659

### Max Norm

In [31]:
x

array([25,  3,  5])

In [33]:
np.max([np.abs(x[0]), np.abs(x[1]), np.abs(x[2])])

25

### Base Vector, Orthogonal Vectors, Orthonormal Vectors

In [34]:
i = np.array([0,1])

In [35]:
j = np.array([1,0])

In [36]:
np.dot(i,j)

0

### Matrix 

In [37]:
X = np.array ([[2,3,4],[4,5,6],[8,6,7]])
X

array([[2, 3, 4],
       [4, 5, 6],
       [8, 6, 7]])

In [38]:
X.shape

(3, 3)

In [39]:
X.size

9

In [43]:
X[2,:]

array([8, 6, 7])

In [53]:
X[:,1]

array([3, 5, 6])

In [54]:
X[2,2]

7

In [57]:
X[0:3, 3:0]

array([], shape=(3, 0), dtype=int32)

### Higher-Rank Tensors

In [86]:
X = pt.zeros([32,28,28,3])
X

tensor([[[[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         ...,

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]],

         [[0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.],
          ...,
          [0., 0., 0.],
          [0., 0., 0.],
          [0., 0., 0.]]],


        [[[0., 0.

In [78]:
Y = pt.tensor([[[[1,2],[3,5]],[[4,6],[5,6]]],[[[1,2],[3,5]],[[4,6],[5,6]]]])
Y

tensor([[[[1, 2],
          [3, 5]],

         [[4, 6],
          [5, 6]]],


        [[[1, 2],
          [3, 5]],

         [[4, 6],
          [5, 6]]]])

In [92]:
Y.dim()

4

In [121]:
images_tf = tf.zeros([12,3,4,5], dtype = tf.int16)

In [122]:
images_tf

<tf.Tensor: shape=(12, 3, 4, 5), dtype=int16, numpy=
array([[[[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]],

        [[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]],

        [[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]]],


       [[[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]],

        [[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]],

        [[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]]],


       [[[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]],

        [[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0]],

        [[0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0, 0, 0],
         [0, 0, 0

In [126]:
tf.rank(images_tf)

<tf.Tensor: shape=(), dtype=int32, numpy=4>

In [142]:
q1 = pt.tensor([[25,2,-3,-23]])
q1

tensor([[ 25,   2,  -3, -23]])

In [143]:
q1_t = q1.T
q1_t

tensor([[ 25],
        [  2],
        [ -3],
        [-23]])

In [139]:
q2 = tf.Variable([[42,4,7,99],[-99,-3,17,22]])
q2

<tf.Variable 'Variable:0' shape=(2, 4) dtype=int32, numpy=
array([[ 42,   4,   7,  99],
       [-99,  -3,  17,  22]])>

In [140]:
tf.rank(q2)

<tf.Tensor: shape=(), dtype=int32, numpy=2>

In [141]:
q2[1,2]

<tf.Tensor: shape=(), dtype=int32, numpy=17>